### ISSUE TITLE
Http -> Https redirect from TCP ELB offloading SSL not working, 301 redirect loop. Kubernetes/Kops

### ISSUE THREAD

**Issue by JimtotheB** [containous/traefik#3551]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

tldr: AWS ELB in TCP/SSL mode terminating ssl at the ELB, causes a redirect loop with the configs and settings below. An AWS ELB in HTTP/HTTPS mode, with the same settings below works as expected.

Created a AWS ELB loadbalancer in TCP/SSL mode, with ssl offloading on the ELB with the following service.

```yaml
---
kind: Service
apiVersion: v1
metadata:
  name: traefik-ingress-service
  namespace: kube-system
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
    service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "3600"
    service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: '*'
    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: "<snip>
    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: https
spec:
  externalTrafficPolicy: Cluster
  selector:
    app: traefik-ingress-lb
  ports:
    - protocol: TCP
      port: 80
      targetPort: http
      name: http
    - protocol: TCP
      port: 443
      targetPort: http
      name: https
  type: LoadBalancer

```

targeting a Traefik Daemonset

```yaml
---
kind: DaemonSet
apiVersion: extensions/v1beta1
metadata:
  name: traefik-ingress-controller
  namespace: kube-system
  labels:
    app: traefik-ingress-lb
spec:
  template:
    metadata:
      labels:
        app: traefik-ingress-lb
        name: traefik-ingress-lb
    spec:
      serviceAccountName: traefik-ingress-controller
      terminationGracePeriodSeconds: 60
      volumes:
      - name: config
        configMap:
          name: traefik-configuration
      containers:
      - image: traefik:1.6.4-alpine
        name: traefik-ingress-lb
        ports:
        - name: http
          containerPort: 80
          hostPort: 80
        - name: admin
          containerPort: 8080
          hostPort: 8080
        volumeMounts:
        - mountPath: "/config"
          name: "config"
        securityContext:
          capabilities:
            drop:
            - ALL
            add:
            - NET_BIND_SERVICE
        args:
        - --configFile=/config/traefik.toml
        - --api
        - --kubernetes
        - --logLevel=DEBUG
```

with `/config/traefik.toml` consisting of the following

```toml
defaultEntryPoints = ["http"]
[entryPoints]
  [entryPoints.http]
    address = ":80"
    # Enable ProxyProtocol
    [entryPoints.http.proxyProtocol]
      trustedIPs = ["172.31.0.0/16", "100.64.0.0/10"]
```
My backend service has the following annotations:

```yaml
annotations:
    ingress.kubernetes.io/ssl-redirect: "true"
    ingress.kubernetes.io/ssl-proxy-headers: "X-Forwarded-Proto: https"
```

### What did you expect to see?

I expected to see a request to http://example.com redirect to https://example.com

### What did you see instead?

A 301 redirect loop.

### Output of `traefik version`: (_What version of Traefik are you using?_)

traefik:1.6.4-alpine

```
Version:      v1.6.4
Codename:     tetedemoine
Go version:   go1.10.3
Built:        2018-06-15_03:12:50PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
defaultEntryPoints = ["http"]
[entryPoints]
  [entryPoints.http]
    address = ":80"
    # Enable ProxyProtocol
    [entryPoints.http.proxyProtocol]
      trustedIPs = ["172.31.0.0/16", "100.64.0.0/10"]
```

### If applicable, please paste the log output in DEBUG level (`--logLevel=DEBUG` switch)

```
time="2018-07-02T06:08:06Z" level=info msg="Using TOML configuration file /config/traefik.toml"
time="2018-07-02T06:08:06Z" level=info msg="Traefik version v1.6.4 built on 2018-06-15_03:12:50PM"
time="2018-07-02T06:08:06Z" level=info msg="\nStats collection is disabled.\nHelp us improve Traefik by turning this feature on :)\nMore details on: https://docs.traefik.io/basics/#collected-data\n"
time="2018-07-02T06:08:06Z" level=debug msg="Global configuration loaded {\"LifeCycle\":{\"RequestAcceptGraceTimeout\":0,\"GraceTimeOut\":10000000000},\"GraceTimeOut\":0,\"Debug\":false,\"CheckNewVersion\":true,\"SendAnonymousUsage\":fa
lse,\"AccessLogsFile\":\"\",\"AccessLog\":{\"format\":\"common\"},\"TraefikLogsFile\":\"\",\"TraefikLog\":null,\"Tracing\":null,\"LogLevel\":\"DEBUG\",\"EntryPoints\":{\"http\":{\"Address\":\":80\",\"TLS\":null,\"Redirect\":null,\"Auth\
":null,\"WhitelistSourceRange\":null,\"WhiteList\":null,\"Compress\":false,\"ProxyProtocol\":{\"Insecure\":false,\"TrustedIPs\":[\"172.31.0.0/16\",\"100.64.0.0/10\"]},\"ForwardedHeaders\":{\"Insecure\":true,\"TrustedIPs\":null}},\"traef
ik\":{\"Address\":\":8080\",\"TLS\":null,\"Redirect\":null,\"Auth\":null,\"WhitelistSourceRange\":null,\"WhiteList\":null,\"Compress\":false,\"ProxyProtocol\":null,\"ForwardedHeaders\":{\"Insecure\":true,\"TrustedIPs\":null}}},\"Cluster
\":null,\"Constraints\":[],\"ACME\":null,\"DefaultEntryPoints\":[\"http\"],\"ProvidersThrottleDuration\":2000000000,\"MaxIdleConnsPerHost\":200,\"IdleTimeout\":0,\"InsecureSkipVerify\":false,\"RootCAs\":null,\"Retry\":null,\"HealthCheck
\":{\"Interval\":30000000000},\"RespondingTimeouts\":null,\"ForwardingTimeouts\":null,\"AllowMinWeightZero\":false,\"Web\":null,\"Docker\":null,\"File\":null,\"Marathon\":null,\"Consul\":null,\"ConsulCatalog\":null,\"Etcd\":null,\"Zooke
eper\":null,\"Boltdb\":null,\"Kubernetes\":{\"Watch\":true,\"Filename\":\"\",\"Constraints\":[],\"Trace\":false,\"TemplateVersion\":0,\"DebugLogGeneratedTemplate\":false,\"Endpoint\":\"\",\"Token\":\"\",\"CertAuthFilePath\":\"\",\"Disab
lePassHostHeaders\":false,\"EnablePassTLSCert\":false,\"Namespaces\":null,\"LabelSelector\":\"\",\"IngressClass\":\"\"},\"Mesos\":null,\"Eureka\":null,\"ECS\":null,\"Rancher\":null,\"DynamoDB\":null,\"ServiceFabric\":null,\"Rest\":null,
\"API\":{\"EntryPoint\":\"traefik\",\"Dashboard\":true,\"Debug\":false,\"CurrentConfigurations\":null,\"Statistics\":null},\"Metrics\":null,\"Ping\":null}"
time="2018-07-02T06:08:06Z" level=info msg="Preparing server http &{Address::80 TLS:<nil> Redirect:<nil> Auth:<nil> WhitelistSourceRange:[] WhiteList:<nil> Compress:false ProxyProtocol:0xc420506a60 ForwardedHeaders:0xc4203876e0} with re
adTimeout=0s writeTimeout=0s idleTimeout=3m0s"
time="2018-07-02T06:08:06Z" level=info msg="Enabling ProxyProtocol for trusted IPs [172.31.0.0/16 100.64.0.0/10]"
time="2018-07-02T06:08:06Z" level=info msg="Preparing server traefik &{Address::8080 TLS:<nil> Redirect:<nil> Auth:<nil> WhitelistSourceRange:[] WhiteList:<nil> Compress:false ProxyProtocol:<nil> ForwardedHeaders:0xc420387740} with read
Timeout=0s writeTimeout=0s idleTimeout=3m0s"
time="2018-07-02T06:08:06Z" level=info msg="Starting provider *kubernetes.Provider {\"Watch\":true,\"Filename\":\"\",\"Constraints\":[],\"Trace\":false,\"TemplateVersion\":0,\"DebugLogGeneratedTemplate\":false,\"Endpoint\":\"\",\"Token\
":\"\",\"CertAuthFilePath\":\"\",\"DisablePassHostHeaders\":false,\"EnablePassTLSCert\":false,\"Namespaces\":null,\"LabelSelector\":\"\",\"IngressClass\":\"\"}"
time="2018-07-02T06:08:06Z" level=info msg="Starting server on :80"
time="2018-07-02T06:08:06Z" level=info msg="Starting server on :8080"
time="2018-07-02T06:08:06Z" level=debug msg="Using Ingress label selector: \"\""
time="2018-07-02T06:08:06Z" level=info msg="ingress label selector is: \"\""
time="2018-07-02T06:08:06Z" level=info msg="Creating in-cluster Provider client"
time="2018-07-02T06:08:07Z" level=debug msg="Received Kubernetes event kind *v1beta1.Ingress"
time="2018-07-02T06:08:07Z" level=debug msg="Received Kubernetes event kind *v1.Secret"
time="2018-07-02T06:08:07Z" level=debug msg="Skipping Kubernetes event kind *v1.Secret"
time="2018-07-02T06:08:07Z" level=debug msg="Received Kubernetes event kind *v1.Secret"
time="2018-07-02T06:08:07Z" level=debug msg="Skipping Kubernetes event kind *v1.Secret"
time="2018-07-02T06:08:07Z" level=debug msg="Configuration received from provider kubernetes: {\"backends\":{\"snip/\":{\"servers\":{\"snip\":{\"url\":\"http://100.96.5.58:8080\",\"weight\":1}
},\"loadBalancer\":{\"method\":\"wrr\"}}},\"frontends\":{\"snip/\":{\"entryPoints\":[\"http\"],\"backend\":\"snip/\",\"routes\":{\"/\":{\"rule\":\"PathPrefix:/\"},\"snip\":{\"r
ule\":\"Host:snip\"}},\"passHostHeader\":true,\"priority\":0,\"basicAuth\":[],\"headers\":{\"sslRedirect\":true,\"sslProxyHeaders\":{\"X-Forwarded-Proto\":\"https\"}}}}}"
time="2018-07-02T06:08:07Z" level=debug msg="Creating frontend snip/"
time="2018-07-02T06:08:07Z" level=debug msg="Wiring frontend snip/ to entryPoint http"
time="2018-07-02T06:08:07Z" level=debug msg="Creating route / PathPrefix:/"
time="2018-07-02T06:08:07Z" level=debug msg="Creating route snip Host:snip"
time="2018-07-02T06:08:07Z" level=debug msg="Creating backend snip/"
time="2018-07-02T06:08:07Z" level=debug msg="Creating load-balancer wrr"
time="2018-07-02T06:08:07Z" level=debug msg="Creating server snip at http://100.96.5.58:8080 with weight 1"
time="2018-07-02T06:08:07Z" level=debug msg="Adding secure middleware for frontend snip/"
time="2018-07-02T06:08:07Z" level=info msg="Server configuration reloaded on :80"
time="2018-07-02T06:08:07Z" level=info msg="Server configuration reloaded on :8080"
```

I have this up live on a publicly accessible cluster and can push whatever changes need to test if anyone has any suggestions.

**Comment by JimtotheB** [comment]
@dduportal websockets 

You are exactly right the X-Forwarded-Proto is indeed stripped. I made a bad assumption here, that because it was present in the response from the echoserver, that it was being set by the ELB, when in reality it is set by nginx on that particular backend.

I guess I am going to try sending port 80/443 traffic from the ELB to different Traefik frontends and just rewrite everything that comes in from port 80 on the ELB.

**Comment by dduportal** [comment]
If you are running Kubernetes 1.9+, you could use a "NLB" (Network Load Balancer) for your ELB, which supports ebsockets and HTTP/2 out of the box.

Here is a great article showing you the right annotation: https://blog.giantswarm.io/load-balancer-service-use-cases-on-aws/#i-want-to-use-the-new-nlb-support-in-kubernetes-19 
I used it when reproducing your case and websocket as working well :)

Hope it helps.

**Comment by JimtotheB** [comment]
I was aware of network load balancers, but avoiding it because I wanted the SSL offload. Ill have to explore getting ACM certs into Traefik or barring that using Lets Encrypt.

**Comment by JimtotheB** [comment]
@dduportal Thanks for the quick help. Are there docs somewhere for running Traefik behind a NLB?

**Comment by dduportal** [comment]
@JimtotheB  I don't know if there is any: at least I was not able to found one.

**Comment by JimtotheB** [comment]
Yeah, I cant get any of the registered instances to pass the health checks. Basically using the configuration above but with the `service.beta.kubernetes.io/aws-load-balancer-type: "nlb"` annotation as the only one.

**Comment by Pehesi97** [comment]
> Yeah, I cant get any of the registered instances to pass the health checks. Basically using the configuration above but with the `service.beta.kubernetes.io/aws-load-balancer-type: "nlb"` annotation as the only one.

Same here.
